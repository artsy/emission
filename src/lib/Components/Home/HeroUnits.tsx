import * as React from "react"
import * as Relay from "react-relay"

import { LayoutChangeEvent, ScrollView, TouchableHighlight, View, ViewProperties } from "react-native"

import SwitchBoard from "../../NativeModules/SwitchBoard"
import OpaqueImageView from "../OpaqueImageView"
import Headline from "../Text/Headline"

interface Props extends ViewProperties, RelayProps {}

interface State {
  width?: number
  height?: number
  margin?: number
  fontSize?: number
}

class HeroUnits extends React.Component<Props, State> {
  constructor(props) {
    super(props)
    this.state = {
      width: null,
      height: null,
      margin: null,
      fontSize: 0,
    }
  }

  handleLayout = (event: LayoutChangeEvent) => {
    const { nativeEvent: { layout: { width } } } = event
    const margin = width > 700 ? 40 : 20
    const height = width > 700 ? 300 : 160
    const fontSize = width > 700 ? 30 : 20
    this.setState({ width, height, margin, fontSize })
  }

  handlePress = heroUnit => {
    SwitchBoard.presentNavigationViewController(this, heroUnit.href)
  }

  renderHeroUnit(heroUnit, margin, width, height) {
    return (
      <TouchableHighlight
        key={heroUnit.__id}
        onPress={() => this.handlePress(heroUnit)}
        underlayColor="black"
        style={{ marginLeft: 5, marginRight: 5, marginTop: 20, width, height }}
      >
        <View style={{ flex: 1, justifyContent: "center" }}>
          <OpaqueImageView
            imageURL={width > 600 ? heroUnit.wide_image_url : heroUnit.narrow_image_url}
            style={{ position: "absolute", top: 0, width, height }}
          />
          <View style={{ position: "absolute", top: 0, width, height, backgroundColor: "black", opacity: 0.3 }} />
          <View style={{ marginLeft: margin, backgroundColor: "transparent" }}>
            <Headline style={{ color: "white", fontSize: 12 }}>{heroUnit.heading}</Headline>
            <Headline style={{ color: "white", fontSize: this.state.fontSize }}>{heroUnit.title}</Headline>
          </View>
        </View>
      </TouchableHighlight>
    )
  }

  renderHeroUnits() {
    const { width, height, margin } = this.state
    if (width && height && margin) {
      const imageWidth = width - margin * 2
      return (
        <ScrollView
          style={{ overflow: "visible", marginLeft: margin - 5, marginRight: margin - 5 }}
          horizontal={true}
          pagingEnabled={true}
          showsHorizontalScrollIndicator={false}
        >
          {(this.props.hero_units || []).map(heroUnit => this.renderHeroUnit(heroUnit, margin, imageWidth, height))}
        </ScrollView>
      )
    }
  }

  render() {
    return <View onLayout={this.handleLayout}>{this.renderHeroUnits()}</View>
  }
}

export default Relay.createContainer(HeroUnits, {
  fragments: {
    hero_units: () => Relay.QL`
      fragment on HomePageHeroUnit @relay(plural: true) {
        __id
        href
        title
        heading
        narrow_image_url: background_image_url(version: NARROW)
        wide_image_url: background_image_url(version: WIDE)
      }
    `,
  },
})

interface RelayProps {
  hero_units: Array<{
    __id: string
    href: string | null
    title: string | null
    heading: string | null
    narrow_image_url: string | null
    wide_image_url: string | null
  } | null> | null
}
