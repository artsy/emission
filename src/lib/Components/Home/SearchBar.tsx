import * as React from "react"

import { Dimensions, Image, StyleSheet, Text, TextStyle, TouchableWithoutFeedback, View, ViewStyle } from "react-native"

import colors from "../../../data/colors"
import Switchboard from "../../native_modules/switch_board"

const sideMargin = Dimensions.get("window").width > 700 ? 40 : 20
const leftInnerMargin = Dimensions.get("window").width > 700 ? 20 : 15

export default class SearchBar extends React.Component<any, any> {
  render() {
    return (
      <TouchableWithoutFeedback onPress={this.handleTap.bind(this)}>
        <View style={styles.container}>
          <Image style={styles.searchIcon} source={require("../../../../images/SearchButton.png")} />
          <Text style={styles.text}>Search for artists and artworks...</Text>
        </View>
      </TouchableWithoutFeedback>
    )
  }

  handleTap() {
    Switchboard.presentModalViewController(this, "/search")
  }
}

interface Styles {
  container: ViewStyle
  text: TextStyle
  searchIcon: ViewStyle
}

const styles = StyleSheet.create<Styles>({
  container: {
    height: 50,
    backgroundColor: colors["gray-light"],
    borderColor: colors["gray-regular"],
    borderWidth: 1,
    marginLeft: sideMargin,
    marginRight: sideMargin,
    marginTop: 10,
    alignItems: "center",
    flexDirection: "row",
  },
  text: {
    fontFamily: "AGaramondPro-Regular",
    fontSize: 16,
    marginTop: 5,
  },
  searchIcon: {
    marginLeft: leftInnerMargin,
    marginRight: 10,
    height: 16,
    width: 16,
  },
})
