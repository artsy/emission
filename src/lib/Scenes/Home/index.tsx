import React from "react"
import { AppState, View } from "react-native"
import ScrollableTabView from "react-native-scrollable-tab-view"
import styled from "styled-components/native"

import { Schema, screenTrack } from "lib/utils/track"

import SwitchBoard from "lib/NativeModules/SwitchBoard"
import { options } from "lib/options"
import { Router } from "lib/utils/router"

import WorksForYou from "lib/Containers/WorksForYou"
import ForYou from "./Components/ForYou"
import Sales from "./Components/Sales"

import { ForYouRenderer, WorksForYouRenderer } from "lib/relay/QueryRenderers"
import { SalesRenderer } from "lib/Scenes/Home/Components/Sales/Relay/SalesRenderer"
import renderWithLoadProgress from "lib/utils/renderWithLoadProgress"

import DarkNavigationButton from "lib/Components/Buttons/DarkNavigationButton"
import TabBar, { Tab } from "lib/Components/TabBar"

const TabBarContainer = styled.View``

interface Props {
  selectedArtist?: string
  selectedTab?: number
  initialTab: number
  tracking: any
  isVisible: boolean
}

interface State {
  appState: string
  selectedTab: number
}

const ArtistsWorksForYouTab = 0
const ForYouTab = 1
const AuctionsTab = 2

const screenSchemaForCurrentTab = currentSelectedTab => {
  let screenType

  if (currentSelectedTab === ArtistsWorksForYouTab) {
    screenType = Schema.PageNames.HomeArtistsWorksForYou
  } else if (currentSelectedTab === ForYouTab) {
    screenType = Schema.PageNames.HomeForYou
  } else if (currentSelectedTab === AuctionsTab) {
    screenType = Schema.PageNames.HomeAuctions
  }

  return screenType
}

// This kills two birds with one stone:
// It's necessary to wrap all tracks nested in this component, so they dispatch properly
// Also, it'll only fire when the home screen is mounted, the only event we would otherwise miss with our own callbacks
@screenTrack((props: Props) => ({
  context_screen: screenSchemaForCurrentTab(props.initialTab),
  context_screen_owner_type: null,
}))
export default class Home extends React.Component<Props, State> {
  tabView?: ScrollableTabView | any

  constructor(props) {
    super(props)

    this.state = {
      appState: AppState.currentState,
      selectedTab: this.props.initialTab,
    }
  }

  componentDidMount() {
    AppState.addEventListener("change", this._handleAppStateChange)
  }

  componentWillUnmount() {
    AppState.removeEventListener("change", this._handleAppStateChange)
  }

  // FIXME: Make a proper "viewDidAppear" callback / event emitter
  // https://github.com/artsy/emission/issues/930
  // This is called when the overall home component appears in Eigen
  // We use it to dispatch screen events at that point
  componentWillReceiveProps(newProps) {
    // Only if we weren't visible, but we are now, should we fire analytics for this.
    if (newProps.isVisible && !this.props.isVisible) {
      this.fireHomeScreenViewAnalytics()
    }
    if (this.props.selectedTab !== newProps.selectedTab) {
      this.setState({ selectedTab: newProps.selectedTab }, () => this.tabView.goToPage(this.state.selectedTab))
    }
  }

  _handleAppStateChange = nextAppState => {
    // If we are coming back to the app from the background with a selected artist, make sure we're on the Artists tab
    if (this.props.selectedArtist && this.state.appState.match(/inactive|background/) && nextAppState === "active") {
      this.tabView.goToPage(ArtistsWorksForYouTab)
    }
    this.setState({ appState: nextAppState })
  }

  render() {
    // This option doesn't exist today, so it will be undefined, it can
    // be added to Echo's set of features if we want to disable it.
    const showConsignmentsSash = !options.hideConsignmentsSash
    return (
      <View style={{ flex: 1 }}>
        <ScrollableTabView
          initialPage={this.props.initialTab || ArtistsWorksForYouTab}
          ref={tabView => (this.tabView = tabView)}
          onChangeTab={selectedTab => this.setSelectedTab(selectedTab)}
          renderTabBar={props => (
            <TabBarContainer>
              <TabBar {...props} />
            </TabBarContainer>
          )}
        >
          {/* FIXME:
      A thin space has been added in front of the tab label names to compensate for trailing space added by the
      wider letter-spacing. Going forward, this would ideally be dealt with through letter indentation. */}
          <Tab tabLabel=" Artists">
            <WorksForYouRenderer
              render={renderWithLoadProgress(WorksForYou)}
              selectedArtist={this.props.selectedArtist}
            />
          </Tab>
          <Tab tabLabel=" For You">
            <ForYouRenderer render={renderWithLoadProgress(ForYou)} />
          </Tab>
          <Tab tabLabel=" Auctions">
            <SalesRenderer render={renderWithLoadProgress(Sales)} />
          </Tab>
        </ScrollableTabView>
        {showConsignmentsSash && (
          <DarkNavigationButton
            title="Sell works from your collection through Artsy"
            onPress={this.openLink.bind(this)}
          />
        )}
      </View>
    )
  }

  openLink() {
    SwitchBoard.presentNavigationViewController(this, Router.ConsignmentsStartSubmission)
  }

  setSelectedTab(selectedTab) {
    this.setState({ selectedTab: selectedTab.i }, this.fireHomeScreenViewAnalytics)
  }

  fireHomeScreenViewAnalytics = () => {
    this.props.tracking.trackEvent({
      context_screen: screenSchemaForCurrentTab(this.state.selectedTab),
      context_screen_owner_type: null,
    })
  }
}
