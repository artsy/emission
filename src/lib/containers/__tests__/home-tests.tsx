import * as React from 'react'
import * as renderer from 'react-test-renderer'

import Home from '../home'

// jest.mock('../../components/home/artist_rails/artist_rail.tsx', () => 'ArtistRail')
// jest.mock('../../components/home/artwork_rails/artwork_rail.tsx', () => 'ArtworkRail')
// jest.mock('../../components/spinner.tsx', () => 'ARSpinner')
// jest.mock('../../components/opaque_image_view.tsx', () => 'AROpaqueImageView')

describe('upon initialization', () => {
  it('starts with isRefreshing set to false', () => {
    const homeComponent = new Home(homeProps())
    expect(homeComponent.state.isRefreshing).toBeFalsy()
  })

  it('creates a data source with a search bar and hero unit', () => {
    const homeComponent = new Home(homeProps())
    const dataSourceBlob = extractListViewDataBlob(homeComponent)
    const searchBar = { type: 'search_bar', data: null }
    const heroUnits = { type: 'hero_units', data: [] }
    expect(dataSourceBlob).toContainEqual(searchBar)
    expect(dataSourceBlob).toContainEqual(heroUnits)
  })

  it('orders modules correctly', () => {
    const homeComponent = new Home(homeProps(3,3,3))
    const dataSourceBlob = extractListViewDataBlob(homeComponent)
    // order should be search bar, hero units, 2 artwork modules, one artist module
    expect(dataSourceBlob[0].type).toBe('search_bar')
    expect(dataSourceBlob[1].type).toBe('hero_units')
    expect(dataSourceBlob[2].type).toBe('artwork')
    expect(dataSourceBlob[3].type).toBe('artwork')
    expect(dataSourceBlob[4].type).toBe('artist')
  })
})

describe('layout', () => {
  it('renders as expected', () => {
    const homeComponent = renderer.create(<Home home={homeProps(3,3,3).home}/>).toJSON()
    expect(homeComponent).toMatchSnapshot()
  })
})

var homeProps = (nHeroUnits?: number, nArtistModules?: number, nArtworkModules?: number) => {
  return {
      home: {
        hero_units: nHeroUnits ? modulesOfType('hero_units', nHeroUnits) : [],
        artist_modules: nArtistModules ? modulesOfType('artist', nArtistModules) : [],
        artwork_modules: nArtworkModules ? modulesOfType('artwork', nArtworkModules) : [],
        }
    }
}

var modulesOfType = (type: string, count: number) => {
  const modules = []
  for (let i = 0; i < count; i++) {
    modules.push({ title: 'Title', type: type })
  }
  return modules
}

var extractListViewDataBlob = (home: Home) => {
  return home.state.dataSource._dataBlob.s1
}
