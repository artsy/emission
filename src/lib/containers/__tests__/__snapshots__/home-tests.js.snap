exports[`layout renders as expected 1`] = `
<ScrollView
  automaticallyAdjustContentInsets={false}
  dataSource={
    ListViewDataSource {
      "items": 6,
    }
  }
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  refreshControl={
    <RefreshControl
      onRefresh={[Function]}
      refreshing={false} />
  }
  removeClippedSubviews={true}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "marginTop": 20,
      "overflow": "visible",
    }
  }>
  <TouchableWithoutFeedback
    onPress={[Function]}>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#f8f8f8",
          "borderColor": "#e5e5e5",
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 50,
          "marginLeft": 20,
          "marginRight": 20,
          "marginTop": 10,
        }
      }>
      <Image
        source={1}
        style={
          Object {
            "height": 16,
            "marginLeft": 15,
            "marginRight": 10,
            "width": 16,
          }
        } />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontFamily": "AGaramondPro-Regular",
            "fontSize": 16,
            "marginTop": 5,
          }
        }>
        Search for artists and artworks...
      </Text>
    </View>
  </TouchableWithoutFeedback>
  <View
    onLayout={[Function]} />
  <ArtworkRail
    rail={
      Object {
        "title": "Title",
        "type": "artwork",
      }
    } />
  <ArtworkRail
    rail={
      Object {
        "title": "Title",
        "type": "artwork",
      }
    } />
  <ArtistRail
    rail={
      Object {
        "title": "Title",
        "type": "artist",
      }
    } />
  <ArtworkRail
    rail={
      Object {
        "title": "Title",
        "type": "artwork",
      }
    } />
</ScrollView>
`;
